Add debug action to kubernetes-worker

Cynerva opened the pull request:
This is dependent on [layer-debug](https://github.com/charms/layer-debug)

* [Example run](https://gist.github.com/Cynerva/0e784ed504d91b7844264bcd691d72b3)
* [Resulting archive](https://dl.dropboxusercontent.com/u/32306657/sos-20161108221709.tar.gz)

Issue https://github.com/juju-solutions/bundle-canonical-kubernetes/issues/76

@wwwtyro @chuckbutler @mbruzek @marcoceppi Feedback plz

chuckbutler commented on cluster/juju/layers/kubernetes-worker/sos-scripts/kubectl:
```diff
@@ -0,0 +1,11 @@
+sudo -iu ubuntu kubectl cluster-info > $SOS_SCRIPT_DIR/cluster-info
+sudo -iu ubuntu kubectl cluster-info dump > $SOS_SCRIPT_DIR/cluster-info-dump
+for obj in pods svc ingress secrets pv pvc rc; do
+  sudo -iu ubuntu kubectl describe $obj --all-namespaces > $SOS_SCRIPT_DIR/describe-$obj
```
wont this ovewrite the file with each object type, and should instead be touched then appended to? `>>` instead of '>'

Cynerva replied:
It's a different file for each object type, `describe-$obj`, so they shouldn't overwrite eachother. Here was the resulting tree after running it:

```
├── kubectl
│   ├── cluster-info
│   ├── cluster-info-dump
│   ├── describe-ingress
│   ├── describe-nodes
│   ├── describe-pods
│   ├── describe-pv
│   ├── describe-pvc
│   ├── describe-rc
│   ├── describe-secrets
│   ├── describe-svc
│   ├── stderr
│   └── stdout
```

I think it's fine - or did I misinterpret what you're saying?

chuckbutler replied:
Nope, cleared that up :)  I wasn't comprehending what was happening in that block and this resolves my concern. 👍

chuckbutler commented:
I can tell its early work, but there's some really good stuff in here. I called out a minor quibble inline.

marcoceppi commented:
Let's put layer-sos in GitHub/charms when it's time to promote

On Wed, Nov 9, 2016, 9:52 AM Charles Butler <notifications@github.com>
wrote:

> I can tell its early work, but there's some really good stuff in here. I
> called out a minor quibble inline.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/juju-solutions/kubernetes/pull/81#issuecomment-259478747>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAET1SNad0KPq-De466MAV-gydmiklQCks5q8ghFgaJpZM4Ks_MT>
> .
>

Cynerva commented:
Renamed sos -> debug. Untested with that change, needs another test run

Cynerva commented:
Gave it another spin, everything looks good on my end.

This is ready for full review + merge if everything checks out.

Do we need to get layer-debug added to [interfaces.juju.solutions](http://interfaces.juju.solutions/) before we merge this?

chuckbutler commented:
I would like to have it officially listed before we push this out for end users to build, please. This will yield broken attempts to build for anyone following along at home otherwise.

chuckbutler commented on cluster/juju/layers/kubernetes-worker/debug-scripts/kubectl:
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
+set -ux
+
+sudo -iu ubuntu kubectl cluster-info > $DEBUG_SCRIPT_DIR/cluster-info
```
Is there any reason these are being run as the ubuntu user? I'm not opposed to this, just curious why

Cynerva replied:
I don't recall why exactly, but kubectl wasn't picking up kubeconfig as root. That problem only occurred within the script, though - microbot etc still worked fine. I'll revisit this.

Cynerva commented:
> * Dump the docker daemon info docker info
> * Dump the currently running container list for contrast/compare against the kubernetes pods
Seems like we're missing networking info, but that is likely to change very very soon.

Ideally those should land in layer-docker at some point, but I can start 'em out here in kubernetes-worker if we're more comfortable with that

> * Seems like we're missing networking info, but that is likely to change very very soon.

Good point. I'll add this to the base layer.

Cynerva commented:
Pushed some changes here:
* Added docker debug script
* Made kubectl debug script a little clearer by passing in kubeconfig instead of running as ubuntu

In addition, some changes landed in layer-debug:
* Added to [interfaces.juju.solutions](http://interfaces.juju.solutions/)
* Added [network](https://github.com/charms/layer-debug/pull/7) debug script

Need +1 from both @chuckbutler and @mbruzek

chuckbutler commented:
I dont think its necessary to do this *right now* but should we add the docker debug business to `layer-docker` so its inherently available in anything being built from `layer-docker`?

marcoceppi commented:
+1

On Fri, Nov 18, 2016 at 11:11 AM Charles Butler <notifications@github.com>
wrote:

> I dont think its necessary to do this *right now* but should we add the
> docker debug business to layer-docker so its inherently available in
> anything being built from layer-docker?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/juju-solutions/kubernetes/pull/81#issuecomment-261570608>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAET1SNWEVz9B73KbV9EiQA2fOkapCzvks5q_c5KgaJpZM4Ks_MT>
> .
>

chuckbutler commented:
+1 LGTM

wwwtyro commented:
ship it
